db/version_set.cc

```diff
@@ -4011,6 +4026,18 @@ Status VersionSet::ProcessManifestWrites(
       size_t idx = 0;
 #endif
       for (auto& e : batch_edits) {
+        if (e->IsWalAddition()) {
+          s = wals_.AddWals(e->GetWalAdditions());
+          if (!s.ok()) {
+            break;
+          }
+        } else if (e->IsWalDeletion()) {
+          s = wals_.DeleteWals(e->GetWalDeletions());
+          if (!s.ok()) {
+            break;
+          }
+        }
```

Comment:
I think this block of applying `e` to `wals_` should occur after writing and syncing MANIFEST succeeds.

Reply:
`AddWal` and `DeleteWal` will do some internal consistency check for the edit.
Even if `AddWal`/`DeleteWal` succeeds, and writing to MANIFEST fails, since it only modifis the in-memory state, it should be fine.
What's your concern here?

Reply:
> `AddWal` and `DeleteWal` will do some internal consistency check for the edit.
> Even if `AddWal`/`DeleteWal` succeeds, and writing to MANIFEST fails, since it only modifis the in-memory state, it should be fine.
> What's your concern here?

You are accessing `wals_` without dbmutex here, which violates the contract.
Consequently, another thread can access `wals_` concurrently before this thread writes to MANIFEST. Should writing to MANIFEST fails, the reader of `wals_` may have already done something by mistake using the unpersisted `wals_`. We should avoid this because it makes recovery hard to reason.
Also note the function's name is "log and apply".

Reply:
Make sense. The check of `AddWal` and `DeleteWal` are more for consistency checking during recovery, if the checks fail here, it's a programming error. So I'll update `wals_` after MANIFEST write.